### PR TITLE
Support tar files for multi-arch images

### LINKED ISSUE #112: oci_tarball: handle OCI-format images
Similar in nature to #74, but when using `rules_oci-v0.3.7`, `oci_tarball` does not seem to work with `oci_image_index` even though that and the `oci_image` rules both say they produce an image.  The `tarball.sh.tpl` script fails because there is no `.config.digest` member in the manifest generated from a `oci_image_index` rule. Just wasn't implemented yet? Or is my issue user error? See `hello_world_linux_image_tar` target vs the individual `hello_world_{}_image_tar` ones that work in the example below. I did not see an option to have `oci_image_index` use a `Docker Manfiest List` format.

<details>
<summary>generated JSON manifest</summary>

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 1080,
      "digest": "sha256:b82d8a82f2b419203fd22883acd9cd64423596746381a662e40554063b1ac194",
      "platform": {
        "os": "linux",
        "architecture": "amd64"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 1080,
      "digest": "sha256:0355d5bbc1c6d6392d796ccc240811ad4656bf04b54021641a692069b5ee250a",
      "platform": {
        "os": "linux",
        "architecture": "amd64"
      }
    }
  ]
}
```

</details>

<details>
<summary>WORKSPACE</summary>

```starlark
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")

http_archive(
    name = "rules_license",
    sha256 = "6157e1e68378532d0241ecd15d3c45f6e5cfd98fc10846045509fb2a7cc9e381",
    urls = [
        "https://github.com/bazelbuild/rules_license/releases/download/0.0.4/rules_license-0.0.4.tar.gz",
        "https://mirror.bazel.build/github.com/bazelbuild/rules_license/releases/download/0.0.4/rules_license-0.0.4.tar.gz",
    ],
)

http_archive(
    name = "platforms",
    sha256 = "5308fc1d8865406a49427ba24a9ab53087f17f5266a7aabbfc28823f3916e1ca",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/platforms/releases/download/0.0.6/platforms-0.0.6.tar.gz",
        "https://github.com/bazelbuild/platforms/releases/download/0.0.6/platforms-0.0.6.tar.gz",
    ],
)

http_archive(
    name = "bazel_skylib",
    sha256 = "b8a1527901774180afc798aeb28c4634bdccf19c4d98e7bdd1ce79d1fe9aaad7",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/bazel-skylib/releases/download/1.4.1/bazel-skylib-1.4.1.tar.gz",
        "https://github.com/bazelbuild/bazel-skylib/releases/download/1.4.1/bazel-skylib-1.4.1.tar.gz",
    ],
)

load("@bazel_skylib//:workspace.bzl", "bazel_skylib_workspace")

bazel_skylib_workspace()

http_archive(
    name = "aspect_bazel_lib",
    sha256 = "ee95bbc80f9ca219b93a8cc49fa19a2d4aa8649ddc9024f46abcdd33935753ca",
    strip_prefix = "bazel-lib-1.29.2",
    url = "https://github.com/aspect-build/bazel-lib/releases/download/v1.29.2/bazel-lib-v1.29.2.tar.gz",
)

load("@aspect_bazel_lib//lib:repositories.bzl", "aspect_bazel_lib_dependencies", "register_jq_toolchains", "register_yq_toolchains")

aspect_bazel_lib_dependencies()

register_jq_toolchains()

register_yq_toolchains()

http_archive(
    name = "rules_cc",
    sha256 = "3d9e271e2876ba42e114c9b9bc51454e379cbf0ec9ef9d40e2ae4cec61a31b40",
    strip_prefix = "rules_cc-0.0.6",
    urls = ["https://github.com/bazelbuild/rules_cc/releases/download/0.0.6/rules_cc-0.0.6.tar.gz"],
)

http_archive(
    name = "rules_pkg",
    sha256 = "8c20f74bca25d2d442b327ae26768c02cf3c99e93fad0381f32be9aab1967675",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/rules_pkg/releases/download/0.8.1/rules_pkg-0.8.1.tar.gz",
        "https://github.com/bazelbuild/rules_pkg/releases/download/0.8.1/rules_pkg-0.8.1.tar.gz",
    ],
)

load("@rules_pkg//:deps.bzl", "rules_pkg_dependencies")

rules_pkg_dependencies()

git_repository(
    name = "aspect_gcc_toolchain",
    commit = "4bd1f94536ee92b7c49673931773038d923ee86e",
    remote = "https://github.com/aspect-build/gcc-toolchain",
)

load("@aspect_gcc_toolchain//toolchain:repositories.bzl", "gcc_toolchain_dependencies")

gcc_toolchain_dependencies()

load("@aspect_gcc_toolchain//toolchain:defs.bzl", "ARCHS", "gcc_register_toolchain")

gcc_register_toolchain(
    name = "gcc_toolchain_aarch64",
    target_arch = ARCHS.aarch64,
)

gcc_register_toolchain(
    name = "gcc_toolchain_armv7",
    target_arch = ARCHS.armv7,
)

gcc_register_toolchain(
    name = "gcc_toolchain_x86_64",
    sysroot_variant = "x86_64-X11",
    target_arch = ARCHS.x86_64,
)

http_archive(
    name = "rules_oci",
    sha256 = "48642588e91e992772b94de06234da6601854fda0ee32a91ce8ef303cf5e5837",
    strip_prefix = "rules_oci-0.3.7",
    url = "https://github.com/bazel-contrib/rules_oci/releases/download/v0.3.7/rules_oci-v0.3.7.tar.gz",
)

load("@rules_oci//oci:dependencies.bzl", "rules_oci_dependencies")

rules_oci_dependencies()

load("@rules_oci//oci:repositories.bzl", "LATEST_CRANE_VERSION", "oci_register_toolchains")

oci_register_toolchains(
    name = "oci",
    crane_version = LATEST_CRANE_VERSION,
)

load("@rules_oci//oci:pull.bzl", "oci_pull")

oci_pull(
    name = "distroless_static",
    digest = "sha256:c3c3d0230d487c0ad3a0d87ad03ee02ea2ff0b3dcce91ca06a1019e07de05f12",
    image = "gcr.io/distroless/static",
    platforms = [
        "linux/amd64",
        "linux/arm64",
    ],
)

oci_pull(
    name = "distroless_cc",
    digest = "sha256:f252d3ca44b7f2c718c67c2020feee7b7fd4ee6bff8a192dfea6e599b7d820ad",
    # TODO(https://github.com/bazel-contrib/rules_oci/issues/74) Use latest images
    # digest = "sha256:fb402c45f3ef485ccd56ca2af2a58615fc47c4978bb3004e8663a83456791f48",
    image = "gcr.io/distroless/cc",
    platforms = [
        "linux/amd64",
        "linux/arm64",
    ],
)
```

</details>

<details>
<summary>BUILD.bazel</summary>

```starlark
load("@aspect_bazel_lib//lib:transitions.bzl", "platform_transition_filegroup")
load("@rules_pkg//pkg:tar.bzl", "pkg_tar")
load("@rules_oci//oci:defs.bzl", "oci_image", "oci_image_index", "oci_tarball")

cc_library(
    name = "greeting",
    srcs = ["greeting.cc"],
    hdrs = ["greeting.h"],
)

cc_binary(
    name = "hello_world",
    srcs = ["hello_world.cpp"],
    features = ["static_libstdcxx"],
    deps = [
        ":greeting",
    ],
)

ARCHS = [
    "linux_x86_64",
    "linux_aarch64",
]

[platform_transition_filegroup(
    name = "hello_world_{}".format(platform),
    srcs = [":hello_world"],
    target_platform = "//platforms:{}".format(platform),
) for platform in ARCHS]

[pkg_tar(
    name = "hello_world_{}_tar".format(platform),
    srcs = [":hello_world_{}".format(platform)],
    include_runfiles = True,
    strip_prefix = "/",
) for platform in ARCHS]

[oci_image(
    name = "hello_world_{}_image".format(platform),
    base = "@distroless_cc",
    entrypoint = ["/examples/cpp/hello_world/hello_world"],
    tars = [
        ":hello_world_{}_tar".format(platform),
    ],
) for platform in ARCHS]

[oci_tarball(
    name = "hello_world_{}_image_tar".format(platform),
    image = ":hello_world_{}_image".format(platform),
    repotags = ["examples/cpp/hello_world_{}:latest".format(platform)],
) for platform in ARCHS]

oci_image_index(
    name = "hello_world_linux_image",
    images = [
        ":hello_world_{}_image".format(platform)
        for platform in ARCHS
    ],
)

oci_tarball(
    name = "hello_world_linux_image_tar",
    image = ":hello_world_linux_image",
    repotags = ["examples/cpp/hello_world_linux:latest"],
)
```

</details>

<details>
<summary>Platform-specific (works)</summary>

```shell
vscode ➜ /workspaces/mycode (main ✗) $ bazel build //examples/cpp/hello_world:hello_world_linux_x86_64_image_tar
INFO: Analyzed target //examples/cpp/hello_world:hello_world_linux_x86_64_image_tar (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //examples/cpp/hello_world:hello_world_linux_x86_64_image_tar up-to-date:
  bazel-bin/examples/cpp/hello_world/hello_world_linux_x86_64_image_tar/tarball.tar
INFO: Elapsed time: 0.196s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
vscode ➜ /workspaces/mycode (main ✗) $ docker load <bazel-bin/examples/cpp/hello_world/hello_world_linux_x86_64_image_tar/tarball.tar
Loaded image: examples/cpp/hello_world_linux_x86_64:latest
vscode ➜ /workspaces/mycode (main ✗) $ docker run --rm examples/cpp/hello_world_linux_x86_64:latest
Hello, world!
```

</details>

<details>
<summary>Multi-arch (doesn't work)</summary>

```shell
vscode ➜ /workspaces/mycode (main ✗) $ bazel build //examples/cpp/hello_world:hello_world_linux_image_tar
INFO: Analyzed target //examples/cpp/hello_world:hello_world_linux_image_tar (3 packages loaded, 6183 targets configured).
INFO: Found 1 target...
ERROR: /workspaces/mycode/examples/cpp/hello_world/BUILD.bazel:61:12: OCI Tarball //examples/cpp/hello_world:hello_world_linux_image_tar failed: (Exit 1): tarball.sh failed: error executing command (from target //examples/cpp/hello_world:hello_world_linux_image_tar) bazel-out/k8-fastbuild/bin/examples/cpp/hello_world/hello_world_linux_image_tar/tarball.sh

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Error: cannot substitute with !!null, can only substitute strings. Hint: Most often you'll want to use '|=' over '=' for this operation
Target //examples/cpp/hello_world:hello_world_linux_image_tar failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.803s, Critical Path: 0.08s
INFO: 3 processes: 3 internal.
FAILED: Build did NOT complete successfully
```

</details>

### PR DESCRIPTION
**Pull request by illicitonion**
Fixes #112

### CODE REVIEW

File: oci/private/tarball.sh.tpl [hunk 1]
```diff
@@ -5,36 +5,120 @@ readonly STAGING_DIR=$(mktemp -d)
 readonly YQ="{{yq}}"
 readonly IMAGE_DIR="{{image_dir}}"
 readonly BLOBS_DIR="${STAGING_DIR}/blobs"
-readonly TARBALL_PATH="{{tarball_path}}"
+readonly TARBALL_PATH="$(pwd)/{{tarball_path}}"
```
Comment (thesayyn) [inline oci/private/tarball.sh.tpl:8]: could you please explain why this change is needed?

**Comment by thesayyn** [comment]
I am confused about how this fixes `oci_image_index` support for oci_tarball.

File: oci/private/tarball.sh.tpl (hunk 1 above)
Comment (illicitonion) [reply oci/private/tarball.sh.tpl:8]: Because we're now globbing all the files for the `tar` invocation at the bottom (so that we can pick up whichever of `index.json` or `manifest.json`), to avoid needing to go through hoops to strip prefix directories, I figured it'd just be easier to `cd` into `${STAGING_DIR}` before invoking `tar` (rather than using `-C`), but that requires the destination being an absolute path

**Comment by illicitonion** [comment]
> I am confused about how this fixes `oci_image_index` support for oci_tarball.

When the image dir contains an image index, this now bundles things up properly so that `docker load` will work with the tar (assuming you've enabled containerd) - it's the same format as `skopeo copy docker://some:image "oci-archive:oci-archive.tar" --multi-arch all -f v2s2` would output.

The layout of the tar file is an `index.json` which is an index of images (which may have annotations indicating their repo tags), which point at manifests by digest (in the blobs directory), and then a blobs directory containing all of the layers (and manifests) referenced.

I've been using it to take an `oci_image_index` and produce a `.tar` file which can be `docker load`ed, just like `oci_tarball` already supports for `oci_image`.

(My goal is to be able to `docker load` an multi-arch image, as I have systems which then consume these `docker load`-able tars)

**Comment by thesayyn** [comment]
I am more confused now. `oci_image_index` doesn't produce more than one image. it just makes a multi-arch image out of multiple images and then produces a single image. and `docker` style tarballs don't support multi-platform images inside the tarballs.  

Could you please add an example to the examples folder to demonstrate how this works?

> bundles things up properly

Could you elaborate a bit more on this?

**Comment by illicitonion** [comment]
Sorry for not being clear! Hopefully the example will help. I've just pushed an example of a multi-platform image with some Go code built for Linux in both amd64 and arm64.

Assuming you've enabled containerd for a local docker install, the tarball it produces can be `docker load`ed and then `docker.com/example:latest` will be runnable. This will work whether you're on an amd64 or arm64 machine.

If you have an Apple Silicon machine, you can do this:
```console
% bazel build //examples/multi_arch_go:image-multiarch-tar                       
INFO: Invocation ID: 0713b09c-525d-42e3-a825-b732f6dd581e
INFO: Analyzed target //examples/multi_arch_go:image-multiarch-tar (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //examples/multi_arch_go:image-multiarch-tar up-to-date:
  bazel-bin/examples/multi_arch_go/image-multiarch-tar/tarball.tar
INFO: Elapsed time: 0.339s, Critical Path: 0.13s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
% docker load <bazel-bin/examples/multi_arch_go/image-multiarch-tar/tarball.tar
Loaded image: docker.com/example
% docker run -it --rm --platform linux/amd64 docker.com/example 
yo
% docker run -it --rm --platform linux/arm64 docker.com/example 
yo
```

From this you can see this `.tar` was docker load-ed and could be run on both architectures. As far as I can tell, this wasn't possbile using rules_oci before this PR.

**Comment by illicitonion** [comment]
(I think one of the key pieces of information here is that while _docker_ style tarballs don't support multi-platform images, oci style ones do, and docker can be configured to work with oci style tarballs)

**Comment by thesayyn** [comment]
> Assuming you've enabled containerd for a local docker install

is this the case for standard setup?

**Comment by illicitonion** [comment]
> > Assuming you've enabled containerd for a local docker install
> 
> is this the case for standard setup?

With docker desktop, there's a "features in development" flag you can enable: https://docs.docker.com/desktop/containerd/

(I'm not actually using docker desktop, I'm using a different system which understands OCI format images, but it's easy to test in docker desktop :))

I'm assuming this feature is reasonably in scope for rules_oci, given it's about oci-spec formatting rather than specifically docker interoperation.

**Comment by thesayyn** [comment]
> (I think one of the key pieces of information here is that while _docker_ style tarballs don't support multi-platform images, oci style ones do, and docker can be configured to work with oci style tarballs)

ideally, we should not implement hacks. it's the reason why we don't produce `both oci and docker` style tarballs in just one tarball by putting both `manifest.json` and `index.json` into the tarball.  I would assume that containerd just supports oci style tarballs and ignores `manifest.json`.  I don't want our API docs to say `oci_image_index` is supported but doesn't in standard setups.

**Comment by illicitonion** [comment]
> ideally, we should not implement hacks

I wouldn't classify this as a hack... It is unfortunate that we're having to generate the `index.json` file ourselves, but I view it as the same level as unfortunate that we're generating a `manifest.json` file ourselves in this file before this PR...

Would it feel more clean if there were an attribute on `oci_tarball` which let you select whether you were expecting an oci or docker format tar?

I agree that it's confusing that there are two possible output formats (oci and docker), and that `oci_tarball` assumes you're outputting the non-oci format and only works if the inputs are non-oci format... Maybe a good solution here would be to rename `oci_tarball` to `docker_tarball`, and make a new rule named `oci_tarball` which only deals with the `oci` format?

**Comment by illicitonion** [comment]
The more I think about it, the more I think that having an `output_format` attr which can take values "docker" or "oci" makes sense, and that we should just error if we can't translate between the formats correctly.

So the current state of the repo would only support "docker", and would error if you gave it an oci image to tar up (e.g. an `oci_image_index`. With this PR, we could support oci -> oci, but we'd still error on oci -> docker or docker -> oci. In the future we may be able to do translation between the formats, but we wouldn't necessarily ever actually implement these.

**Comment by kriscfoster** [comment]
Hi @illicitonion, thank you very much for working on this, it is also a feature we would really like.

We were just trying it out & we see the following:

```
bazel run examples/multi_arch_go:image-multiarch-tar
...
open /var/lib/docker/tmp/docker-import-3288145814/blobs/json: no such file or directory
```

Any idea why this would be?

**Comment by kriscfoster** [comment]
Hi @illicitonion,

Thank you very much for working on this, it is also a feature we would really like. Just curious, when we were trying it out we saw the following error:

```
bazel run examples/multi_arch_go:image-multiarch-tar
...
open /var/lib/docker/tmp/docker-import-3371548048/blobs/json: no such file or directory
```

Just curious if you know why that would be. Thank you.

**Comment by illicitonion** [comment]
@thesayyn Thanks for the review, and sorry for the _ridiculous_ delay in getting back to you.

I've added the `format` attr as described - for now I've only implemented `oci` support for `image_index` targets, but we could add `image` support as/when needed (I imagine I'll probably add it soon, but wanted to keep the PR manageable).

I've also reworked the PR slightly to minimise the diff - the old codepath now hopefully is more clearly mostly unchanged, and the new codepath is mostly just additions not modifications.

Please let me know if you have more ideas or comments! :)

**Comment by illicitonion** [comment]
> Hi @illicitonion,
> 
> Thank you very much for working on this, it is also a feature we would really like. Just curious, when we were trying it out we saw the following error:
> 
> ```
> bazel run examples/multi_arch_go:image-multiarch-tar
> ...
> open /var/lib/docker/tmp/docker-import-3371548048/blobs/json: no such file or directory
> ```
> 
> Just curious if you know why that would be. Thank you.

@kriscfoster It sounds like you're running docker _without_ `containerd` enabled - if you enable it as per https://docs.docker.com/desktop/containerd/ your `docker run` should work.

**Comment by kriscfoster** [comment]
Thank you @illicitonion, this worked!

**Comment by kriscfoster** [comment]
Do we know what is needed for this to make it to `main`? It's working well for us but we would prefer to use it directly from a `main` release instead of from a fork.

File: oci/private/tarball.sh.tpl (hunk 1 above)
Comment (thesayyn) [reply oci/private/tarball.sh.tpl:9]: I see. Can we revert this change as well? There’s a new hermetic toolchain coming up so we won’t need any of these changes and I’d like to keep it same to make it easier for bisecting in case something breaks.

**Comment by alexeagle** [comment]
Note, this will have big merge conflict against https://github.com/bazel-contrib/rules_oci/pull/385, we'll see who gets in first :)

File: oci/private/tarball.sh.tpl (hunk 1 above)
Comment (illicitonion) [reply oci/private/tarball.sh.tpl]: Done - expanded out the glob below so we don't need to absolutify the path here.

**Comment by illicitonion** [comment]
Hopefully this is ~good to go!

@alexeagle What's your timeline for #385? I'm happy to do either rebase (it should be trivial, just every `cp_f_with_mkdir` becaomes an `add_to_tar` call (and removing the `${STAGING_DIR}` prefix from each second arg), and would love to get both in as soon as we can! :)

File: oci/private/tarball.sh.tpl [hunk 2]
```diff
@@ -1,37 +1,140 @@
...
+
+MEDIA_TYPE=$("${YQ}" eval ".manifests[0].mediaType" "${INDEX_FILE}")
+
+# Check that we know how to generate the output format given the input format.
+# We may expand the supported options here in the future, but for now,
+if [[ "${FORMAT}" != "docker" && "${FORMAT}" != "oci" ]]; then
+  echo >&2 "Unknown format: ${FORMAT}. Only support docker|oci"
+  exit 1
+fi
+if [[ "${FORMAT}" == "oci" && "${MEDIA_TYPE}" != "application/vnd.oci.image.index.v1+json" ]]; then
```
Comment (thesayyn) [inline oci/private/tarball.sh.tpl:37]: Can we also allow docker manifests lists here? i don't see a reason why we should restrict this.

File: oci/private/tarball.sh.tpl [hunk 3]
```diff
@@ -1,37 +1,140 @@
...
+  echo >&2 "Format oci is only supported for image_index targets but saw ${MEDIA_TYPE}"
```
Comment (thesayyn) [inline oci/private/tarball.sh.tpl:38]: nit: oci_image_index

File: oci/private/tarball.sh.tpl [hunk 4]
```diff
@@ -1,37 +1,140 @@
...
-# We can't pass newlines to yq due to https://github.com/mikefarah/yq/issues/1430 and
-# we can't update YQ at the moment because structure_test depends on a specific version:
-# see https://github.com/bazel-contrib/rules_oci/issues/212
-repo_tags="$(tr -d '\r' < "${TAGS_FILE}" | tr '\n' '%')" \
+repo_tags="${REPOTAGS[@]}" \
 config="blobs/${CONFIG_DIGEST}" \
 layers="${LAYERS}" \
 "${YQ}" eval \
-        --null-input '.[0] = {"Config": env(config), "RepoTags": "${repo_tags}" | envsubst | split("%") | map(select(. != "")) , "Layers": env(layers) | map( "blobs/" + . + ".tar.gz") }' \
+        --null-input '.[0] = {"Config": env(config), "RepoTags": "${repo_tags}" | envsubst | split(" ") | map(select(. != "")) , "Layers": env(layers) | map( "blobs/" + . + ".tar.gz") }' \
```
Comment (thesayyn) [inline oci/private/tarball.sh.tpl:137]: this change seems unrelated. can we take that back?

Comment (illicitonion) [reply oci/private/tarball.sh.tpl]: I've done this, but I don't think it's how we should leave this long-term; I see two issues:
1. Now the two codepaths handle repotags differently, which may become a source of drift/bugs in the future.
2. I think the "just have bash split things into an array and use the array" approach seems significantly simpler than "read the file, replace newlines with percents, have a 4-line comment explaining why we're doing so, then split again on perecents".

Happy to merge this with the two ways, and maybe follow up unifying them in the future, if that'd make sense? Or just reverting the "Revert repo_tags change" commit in this PR and accepting needting to update other PRs for the conflict? Whatever works best for y'all.

File: oci/private/tarball.sh.tpl (hunk 2 above)
Comment (illicitonion) [reply oci/private/tarball.sh.tpl]: We can! Now we do!

File: oci/private/tarball.sh.tpl (hunk 4 above)
Comment (illicitonion) [reply oci/private/tarball.sh.tpl]: Prefactor PR: https://github.com/bazel-contrib/rules_oci/pull/401

Comment (illicitonion) [reply oci/private/tarball.sh.tpl]: Merged and rebased - PTAL! Thanks!

**Review by thesayyn — APPROVED** [review]

**Comment by ar3s3ru** [comment]
Sorry to dig up an old issue, @illicitonion since you authored these changes and you seem to have experience in the topic, I've hit the same error as @kriscfoster but through Skaffold (which is using Docker Engine API).

More context: https://github.com/GoogleContainerTools/skaffold/issues/9217

Any idea why this wouldn't work? I would assume it should, since I'm able to `podman load` the image produced by `oci_image_index -> oci_tarball` locally. (I also assume you're using Podman, from your earlier comments)

Is there any way we could add support for multi-arch `oci_tarball` in Skaffold using Docker Engine API's client? Or is there a different client that supports both Docker and OCI format?

**Comment by ar3s3ru** [comment]
Friendly ping @illicitonion, if you have any pointers about the above message ☝🏻 🙏🏻

**Comment by illicitonion** [comment]
Sorry, I don't have any useful context here - it sounds like skaffold needs OCI format support.
